proxy/upgrade.go

```diff
@@ -0,0 +1,195 @@
+package proxy
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// isUpgradeRequest returns true if and only if there is a "Connection"
+// key with the value "Upgrade" in Headers of the given request.
+func isUpgradeRequest(req *http.Request) bool {
```

Comment:
this one concerns me a bit, even if we consider the current implementation an experimental feature. The issue is that it always upgrades on any request that has Connection: upgrade. This may break other use cases. One of the reasons is that the proxy would not execute the response filters on upgrade requests. If it is enough now to support websockets, then it is cool, because then in this function, we could simply check the Upgrade header, in addition to the Connection, if it is set to "websocket", because Skipper hasn't supported websockets sofar, so nothing would get broken. But, e.g. if there is a browser out there that tries to upgrade to e.g. spdy (or h2c, who knows), then the requests coming from these browsers would be upgraded, and the response filters would be omitted for those users. If restricting the upgrade to websockets is not enough, then one possible way to come around this issue for now would be to create a cmd line flag where the experimental feature is explicitly enabled. Then later, when the feature is finalized in a more organic way, it can be enabled by default. Let's discuss it.